internal/checks/latency.go

```diff
@@ -0,0 +1,185 @@
+func NewFailureCounts() *FailureCounts {
+	return &FailureCounts{
+		latencyTooHigh: 0,
+		timeoutOrError: 0,
+	}
+}
+
+type LatencyCheck struct {
+	client              client.EthClient
+	Err                 error
+	clientGetter        client.EthClientGetter
+	metricsContainer    *metrics.Container
+	logger              *zap.Logger
+	upstreamConfig      *conf.UpstreamConfig
+	routingConfig       *conf.RoutingConfig
+	methodFailureCounts map[string]*FailureCounts // RPC method -> FailureCounts
+	lock                sync.RWMutex
+	ShouldRun           bool
+}
+
+func NewLatencyChecker(
+	upstreamConfig *conf.UpstreamConfig,
+	routingConfig *conf.RoutingConfig,
+	clientGetter client.EthClientGetter,
+	metricsContainer *metrics.Container,
+	logger *zap.Logger,
+) types.Checker {
+	c := &LatencyCheck{
+		upstreamConfig:      upstreamConfig,
+		routingConfig:       routingConfig,
+		clientGetter:        clientGetter,
+		metricsContainer:    metricsContainer,
+		logger:              logger,
+		methodFailureCounts: make(map[string]*FailureCounts),
+		ShouldRun:           routingConfig.Errors != nil || routingConfig.Latency != nil,
+	}
+
+	if err := c.Initialize(); err != nil {
+		logger.Error("Error initializing LatencyCheck.", zap.Any("upstreamID", c.upstreamConfig), zap.Error(err))
+	}
+
+	return c
+}
+
+func (c *LatencyCheck) Initialize() error {
+	c.logger.Debug("Initializing LatencyCheck.", zap.Any("config", c.upstreamConfig))
+
+	httpClient, err := c.clientGetter(c.upstreamConfig.HTTPURL, &c.upstreamConfig.BasicAuthConfig, &c.upstreamConfig.RequestHeadersConfig)
+	if err != nil {
+		c.Err = err
+		return c.Err
+	}
+
+	c.client = httpClient
+
+	c.runCheck()
+
+	// TODO(polsar): This check is in both PeerCheck and SyncingCheck implementations, but I don't understand what it's supposed to be doing.
+	// The setup is exactly the same in each case, so which method is not supported if the `isMethodNotSupportedErr` call returns `true`?
+	if isMethodNotSupportedErr(c.Err) {
+		c.logger.Debug("LatencyCheck is not supported by upstream, not running check.", zap.String("upstreamID", c.upstreamConfig.ID))
+
+		c.ShouldRun = false
+	}
+
+	return nil
+}
+
+func (c *LatencyCheck) RunCheck() {
+	if c.client == nil {
+		if err := c.Initialize(); err != nil {
+			c.logger.Error("Error initializing LatencyCheck.", zap.Any("upstreamID", c.upstreamConfig.ID), zap.Error(err))
+			c.metricsContainer.LatencyCheckErrors.WithLabelValues(c.upstreamConfig.ID, c.upstreamConfig.HTTPURL, metrics.HTTPInit).Inc()
+		}
+	}
+
+	if c.ShouldRun {
+		c.runCheck()
+	}
+}
+
+func (c *LatencyCheck) runCheck() {
+	if c.client == nil {
+		return
+	}
+
+	latencyConfig := c.routingConfig.Latency
+	if latencyConfig == nil {
+		// TODO(polsar): We still want to check the latency of LatencyCheckMethod using the top-level latency threshold.
+		return
+	}
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	// Iterate over all (method, latencyThreshold) pairs and launch the check for each in parallel.
```

Comment:
I was envisioning latency and error based healthchecks to be "active" healthcheck rather than "passive". Some explanation on the differences [here](https://api7.ai/blog/health-check-ensures-high-availability).

Benefits I see with using active healthchecks:
  1. Less overall RPC requests
  2. More signal - there are more samples (requests made by the graph-node) to consider and these samples are generally more representative than a hardcoded, dummy request.

Reply:
I have some implementation thoughts written [here](https://www.notion.so/alchemotion/Error-and-latency-based-routing-in-the-Satsuma-node-gateway-3a60d896c5844f82bf90f68570b00608?pvs=4#7e08dd7d309b4a2eb780e179edba769e)